doc/compiler/optimizer/IntroLoopOptimizations.md

```diff
@@ -287,3 +287,80 @@ the outer loop decide where to send it. It will send it to the header of the slo
 a switch table at the header of the outer loop and a switch table at the header of the inner loop. There are cascades of
 switch tables until the desired inner loop is reached. The overhead of these switch tables is less likely executed because
 the slow loop is likely not executed based on the assumption that the original guard in the fast loop will not fail.
+
+## 4. Loop Specializer
+
+[Loop Specializer](https://github.com/eclipse/omr/blob/e3a15a993c8aba80582aa1d6f3071e122acbd4c4/compiler/optimizer/LoopVersioner.hpp#L1061-L1071)
+subclasses Loop Versioner and shares about 90% of the code with Loop Versioner. The reason that Loop Specializer shares
+so much code with Loop Versioner is that the transformation it performs is quite like what Loop Versioner does.
+
+Profiler profiles various things at the run time, for example the frequency of a block, or the values of a specific
+expression, etc. Loop Specializer is based on the value profiling information. We keep track of loop invariant values
+or expressions. If they are always a certain value, the loop can be specialized with respect to that value.
+
+In the following example, `n` is loop invariant. It would be great if we know what the value `n` is. If we know `n` is
+a small value like `10` or `4`, the loop can be unrolled completely. The loop would be eliminated. Even if the loop
+cannot be unrolled completely, knowing the precise bounds of the constant, it allows the optimizer to eliminate the
+range of the value within that loop. Rather than testing if the value is between `0` and `n`, we could test if the value
+is between `0` and `10`, or whatever the value is profiled as.
+
+```
+for (int i=0; i < n; ++) {
+   … // n is loop invariant and the profiled value is 10 or a small value
+}
+```
+The intent of Loop Specializer is to introduce constants into loops as much as possible. It can only be done to the
+expressions that do not change within the loop. `n` is profiled because it is loop invariant. If `n` takes the value
+`10` 100% times when it is profiled, when it comes to the next compilation, we look at the profiling information for
+`n`, and we could check outside the loop: if `n == 10` (the profiled value). Before it goes into the loop, it checks if
+`i < 10` instead of `i < n`. Otherwise, if `n` is not equal to `10`, we go into another version of the loop, the original
+version of the loop that checks `i < n`.
```

Comment:
We should eliminate the second part this paragraph and just write it in code:

```
if (n == 10) {
    for (int i=0; i < 10; ++) {
        … // n == 10 can be constant propagated inside the loop
    }
} else {
    // original loop
}
```

Reply:
Addressed in c545174